### PR TITLE
Add support for `assert_true` and `assert_false` to `RSpec/Rails/MinitestAssertions`

### PR DESCRIPTION
**Pull request by ydah**
related to #1485 

This PR add support for `assert_true` and `assert_false` to `RSpec/Rails/MinitestAssertions`.
______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [x] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 1]
```diff
@@ -22,30 +24,23 @@ module Rails
         #   expect(b).not_to eq(a)
         #   expect(a).to eq(nil)
         #   expect(a).not_to be_empty
+        #   expect(a).to be_truthy
+        #   expect(b).to be_falsey
```
Comment (koic) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:28]: `assert_true` and `assert_false` are methods in Test::Unit and don't exist in Minitest (and ActiveSupport::TestCase):

- Minitest ... https://docs.seattlerb.org/minitest/Minitest/Assertions.html
- Test::Unit ... https://test-unit.github.io/test-unit/ja/Test/Unit/Assertions.html

 Should these be supported in `RSpec/Rails/Minitest`?

If so, `assert_true` should be contrasted with `be(true)`, not `be_truthy`. And `assert_false` should be contrasted with `be(false)`, not `be_falsey`.

File: docs/modules/ROOT/pages/cops_rspec_rails.adoc [hunk 2]
```diff
@@ -265,6 +265,8 @@ assert_not_includes a, b
 refute_equal(a, b)
 assert_nil a
 refute_empty(b)
+assert_true(a)
```
Comment (pirj) [inline docs/modules/ROOT/pages/cops_rspec_rails.adoc:268]: The a/b naming for expected/actual, especially when used interchangeably, us getting confusing. Can we use just one for “actual”?

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 3]
```diff
@@ -22,30 +24,23 @@ module Rails
...
-            refute_includes
-            refute_nil
-            refute_empty
-            refute_match
+          ASSERT_SUFFIXES = %w[
+            equal not_equal instance_of not_instance_of includes not_includes
+            match nil not_nil empty not_empty true false
           ].freeze
+          REFUTE_SUFFIXES = %w[
+            equal instance_of includes nil empty match
```
Comment (pirj) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:39]: This decreases discoverability.
Are we hitting the line limit for this cop? What about grouping assertions per sub-cop (eg one line for “equal” etc)?

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 4]
```diff
@@ -14,6 +14,8 @@ module Rails
         #   refute_equal(a, b)
         #   assert_nil a
         #   refute_empty(b)
+        #   assert_true(a)
+        #   assert_false b
```
Comment (pirj) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: Does it make sense to also detect plain `assert`/`refute` from Minitest?

**Review by pirj — COMMENTED** [review]
Looks mostly good.
I agree with @koic that we should auto-correct to stricter be(true)/be(false).

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 1 above)
Comment (pirj) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:28]: The relationship between test::unit and minitest assertions is not trivial https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/test_unit_assertions_adapter.rb#L9

The cop, despite the name, is taking care of all of those assertions.

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 4 above)
Comment (koic) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: This is a technical note. Minitest's `assert` is compatible with `be_truthy`, and `refute` is compatible with `be_falsey`. This means that the behavior of Test::Unit's `assert_true` / `assert_false` and Minitest's `assert` / `refute` are different.

- Test::Unit ... https://github.com/test-unit/test-unit/blob/3.6.1/lib/test/unit/assertions.rb#L1379
- Minitest ... https://github.com/minitest/minitest/blob/v5.21.2/lib/minitest/assertions.rb#L180

Comment (pirj) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: Thanks for bringing this up!

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 3 above)
Comment (G-Rath) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:39]: I've opened #1790 which goes in another approach to this that should retain the readability

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 5]
```diff
@@ -14,6 +14,8 @@ module Rails
         #   refute_equal(a, b)
         #   assert_nil a
         #   refute_empty(b)
+        #   assert_true(a)
+        #   assert_false(a)
```
Comment (koic) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: I have already commented in detail, these are Test::Unit APIs, not Minitest APIs. It seems to contradict the description of "`Check if using Minitest matchers`".

File: spec/rubocop/cop/rspec/rails/minitest_assertions_spec.rb [hunk 6]
```diff
@@ -506,6 +506,98 @@
         expect(a).not_to be_empty
       RUBY
     end
+
+    it 'registers an offense when using `assert_true`' do
```
Comment (G-Rath) [inline spec/rubocop/cop/rspec/rails/minitest_assertions_spec.rb:510]: this is currently within the context for `*_empty` - ideally each assertion class should have its own `context`

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 5 above)
Comment (pirj) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: We don’t plan to introduce separate cops for teat::unit or rails-specific assertions. It handles all minitest-like assertions.

**Comment by ydah** [comment]
Ah I missed some comments. Hmm. I think we should distinguish between Minitest and Test:: Unit.
I thought it best not to support `assert_true` and `assert_false` in this Cop. It would be better to discuss the extent of this support first. For example, it looks like it could be transferred to a Cop called `RSpec/TestUnitAssertions`.

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb (hunk 5 above)
Comment (koic) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb:18]: I see, it would probably help to prevent confusion by mentioning such distinctions in the documentation.

**Comment by G-Rath** [comment]
fwiw I currently vote for "have a single cop" because these all seem to be using the same structure and as a downstream user my focus is on "use `expect` consistently" - that doesn't matter if the assertions come from Minitest, Test::Unit, or Rails.

Put another way: I don't think I've yet seen a technical reason for having different cops? i.e. the assertions so far have followed the same logic and structure regardless of where they're actually from.

If people want to rename it to something like `RSpec/PreferExpectInsteadOfAssert`, that'd be fine with me but I think having two cops based on "Minitest vs TestUnit" when they have the same behaviours is just adding more work for ourselves (and to take that a step further: should we have a third cop dedicated to Rails' extension assertions?)

**Comment by ydah** [comment]
@G-Rath Thank you for expressing your opinion.

> Put another way: I don't think I've yet seen a technical reason for having different cops?

In many cases, I think there is no problem to combine them into one cop. However, as the devil's advocate, I would like to deepen this discussion.

For example, the following comments seem to be points where different actions are required. The compatibility of `assert` and `refute` is different between Minitest and Test:: Unit.
- https://github.com/rubocop/rubocop-rspec/pull/1786#discussion_r1460740373

Also, the adapter to Minitest is defined in rspec-rails, which is appropriate for the RSpec/Rails department.
https://github.com/rspec/rspec-rails/blob/main/lib/rspec/rails/adapters.rb#L140

Note, however, that Unit:: Test exists in rspec-core.
https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/test_unit_assertions_adapter.rb#L9

In other words, it seems odd to enable Cop for the RSpec/Rails department if you are using Unit:: Test-style assertions and not rspec-rails.

**Comment by G-Rath** [comment]
> However, as the devil's advocate, I would like to deepen this discussion.

Yup fully in support of that, and in doing so sooner rather than later.

>In other words, it seems odd to enable Cop for the RSpec/Rails department if you are using Unit:: Test-style assertions and not rspec-rails

Just want to check that I'm understanding what you mean correctly, which is effectively: it's weird for someone who's using just Unit::Test but not Rails, to enable a _Rails_ cop.

If so then yes I agree with that - I'd wondered why this was considered a Rails cop though you've pointed out it's `rspec-rails` that has the Minitest adapter so I'm guessing that is (or was...) the why.

**Comment by pirj** [comment]
> odd to enable Cop for the RSpec/Rails department if you are using Unit:: Test-style assertions and not rspec-rails.

Fair enough, but in the future when we extract rubocop-rspec_rails, we could also extract Rails-specific assertion detectors from this cop, and make the cop open for extension.
We’ll depend on rubocop-rspec from rubocop-rspec_rails, right?

File: spec/rubocop/cop/rspec/rails/minitest_assertions_spec.rb (hunk 6 above)
Comment (pirj) [reply spec/rubocop/cop/rspec/rails/minitest_assertions_spec.rb:510]:
```suggestion
  end

  context 'with boolean assertions' do
    it 'registers an offense when using `assert_true`' do
```

**Review by pirj — APPROVED** [review]
Thank you!

**Comment by G-Rath** [comment]
Probably best to take this to the issue, but there's also Capybara assertions we could be translating, and I'd like to also include [`aggregate_assertions`](https://github.com/tjwp/aggregate_assertions) as it seems very solid and 1:1 with `aggregate_failures`

**Comment by ydah** [comment]
> We’ll depend on rubocop-rspec from rubocop-rspec_rails, right?
Oh, I see. If we needs to extract to rubocop-rspec_rails, maybe we shouldn't add support for this to the RSpec/Rails department's Cop for now?

**Comment by ydah** [comment]
I update this PR.

File: config/default.yml [hunk 7]
```diff
@@ -1172,7 +1172,7 @@ RSpec/Rails/InferredSpecType:
     views: view
 
 RSpec/Rails/MinitestAssertions:
-  Description: Check if using Minitest matchers.
+  Description: Check if using Minitest like matchers.
```
Comment (pirj) [inline config/default.yml:1175]:
```suggestion
  Description: Check if using Minitest-like matchers.
```

**Review by pirj — APPROVED** [review]
Thank you!

File: lib/rubocop/cop/rspec/rails/minitest_assertions.rb [hunk 8]
```diff
@@ -215,6 +222,46 @@ def assertion
...
+            def_node_matcher 'self.minitest_assertion', <<~PATTERN # rubocop:disable InternalAffairs/NodeMatcherDirective
+              (send nil? {:assert_true} $_ $_?)
+            PATTERN
+
+            def self.match(actual, failure_message)
+              new(nil, actual, failure_message.first)
+            end
+
+            def assertion
+              'eq(true)'
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/rails/minitest_assertions.rb:241]: Just curious – why not `'be(true)'`?

Comment (ydah) [reply lib/rubocop/cop/rspec/rails/minitest_assertions.rb]: Oops, sorry😅 I updated this PR.

**Review by bquorning — APPROVED** [review]
Thank you
